File: Dante1/Module1.07/Zadanie7.17/main.c
#include <stdio.h>
#include <math.h>

double mySin(double x, int n);
double myCos(double x, int n);

double helper (int n)
{
    if (n < 2)
    {
        return 1;
    }
    else
    {
        return n * helper(n - 1);
    }
}

int main()
{
    float kat;

    printf("Enter kat: ");
    int check = scanf("%f",&kat);
    if(check != 1)
    {
        printf("Incorrect input");
        return 1;
    }

    int nSin = 0;
    int nCos = 0;

    while (fabs(mySin(kat, nSin) - sinf(kat)) >= 0.0001)
    {
        nSin++;
    }

    while (fabs(myCos(kat, nCos) - cosf(kat)) >= 0.0001)
    {
        nCos++;
    }

    printf("%i\n%i\n", nSin, nCos);

    return 0;
}

double mySin(double x, int n)
{
    if(n < 0)
    {
        return -1;
    }

    double sum = 0;

    for(int i = 0; i <= n; i++)
    {
        sum = sum + (pow(-1, i) / helper(2 * i + 1)) * pow(x, 2 * i + 1);
    }

    return sum;
}

double myCos(double x, int n)
{
    if(n < 0)
    {
        return -1;
    }

    double sum = 0;

    for(int i = 0; i <= n; i++)
    {
        sum = sum + (pow(-1, i) / helper(2 * i)) * pow(x, 2 * i);
    }

    return sum;
}

File: Dante1/Module1.07/Zadanie7.17/main.rs
fn helper(n: i32) -> f64 {
    if n < 2 {
        return 1.0;
    }
    n as f64 * helper(n - 1)
}

fn my_sin(x: f64, n: i32) -> f64 {
    if n < 0 {
        return -1.0;
    }

    let mut sum: f64 = 0.0;
    for i in 0..=n {
        sum = sum + (f64::powi(-1f64, i) / helper(2 * i + 1)) * f64::powi(x, 2 * i + 1);
    }

    sum
}

fn my_cos(x: f64, n: i32) -> f64 {
    if n < 0 {
        return -1.0;
    }

    let mut sum: f64 = 0.0;
    for i in 0..=n {
        sum = sum + (f64::powi(-1.0, i) / helper(2 * i)) * f64::powi(x, 2 * i);
    }

    sum
}

fn main() {
    let mut input = String::new();

    println!("Enter: ");
    std::io::stdin().read_line(&mut input).expect("Err");
    let kat: f64 = input.trim().parse().expect("Err");

    input.clear();

    println!("Enter: ");
    std::io::stdin().read_line(&mut input).expect("Err");
    let num: i32 = input.trim().parse().expect("Err");

    if num < 0 {
        panic!("Error");
    }

    let mut num1: i32 = 0;
    let mut num2: i32 = 0;

    while f64::abs(my_sin(kat, num1) - f64::sin(kat)) >= 0.0001
    {
        num1 += 1;
    }

    while f64::abs(my_cos(kat, num2) - f64::cos(kat)) >= 0.0001
    {
        num2 += 1;
    }

    println!("{} {}", num1, num2);
}
